File: app/src/main/java/com/qimian233/ztool/service/LogCollectorService.java
package com.qimian233.ztool.service;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.Service;
import android.content.Intent;
import android.os.IBinder;
import android.os.Handler;
import android.os.Looper;

import androidx.core.app.NotificationCompat;

import com.qimian233.ztool.R;

import java.io.*;
import java.text.SimpleDateFormat;
import java.util.*;
import java.util.concurrent.atomic.AtomicBoolean;

/**
 * 应用自身日志采集服务（无需Root权限，通过 PID 过滤 logcat）
 */
public class LogCollectorService extends Service {
    private static final String TAG = "LogCollectorService";
    private static final int NOTIFICATION_ID = 1001;
    private static final String CHANNEL_ID = "log_collector_channel";
    private static final String CHANNEL_NAME = "日志采集服务";

    private static final long MAX_FILE_SIZE = 1024 * 1024; // 1MB
    private static final int MAX_FILES = 20;
    private static final String LOG_DIR = "Log";
    private static final String APP_LOG_SUBDIR = "app";
    private static final String FILE_PREFIX = "app_log_";
    private static final String FILE_SUFFIX = ".log";

    private Process logcatProcess;
    private BufferedWriter currentWriter;
    private File currentFile;
    private final AtomicBoolean isRunning = new AtomicBoolean(false);
    private Thread logcatThread;
    private boolean isRestartMode = false;
    private boolean isForeground = false;
    private Handler mainHandler;
    private NotificationManager notificationManager;

    @Override
    public void onCreate() {
        super.onCreate();
        android.util.Log.d(TAG, "服务 onCreate() 开始");

        mainHandler = new Handler(Looper.getMainLooper());
        notificationManager = getSystemService(NotificationManager.class);

        createNotificationChannel();
        startForegroundImmediately();

        android.util.Log.d(TAG, "服务 onCreate() 完成");
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        android.util.Log.d(TAG, "服务 onStartCommand() 开始");

        if (intent != null) {
            isRestartMode = intent.getBooleanExtra("is_restart", false);
        }

        if (isRestartMode) {
            android.util.Log.d(TAG, "服务重启模式启动");
        }

        if (!isForeground) {
            android.util.Log.w(TAG, "前台服务未启动，立即启动");
            startForegroundImmediately();
        }

        if (!isRunning.get()) {
            isRunning.set(true);
            mainHandler.postDelayed(() -> {
                android.util.Log.d(TAG, "开始启动日志收集");
                startLogCollection();
            }, 100);
        } else {
            android.util.Log.d(TAG, "服务已在运行中，更新通知");
            updateNotification("服务运行中");
        }

        android.util.Log.d(TAG, "服务 onStartCommand() 完成");
        return START_STICKY;
    }

    @Override
    public void onDestroy() {
        android.util.Log.d(TAG, "服务 onDestroy() 开始");
        stopLogCollection();
        android.util.Log.d(TAG, "日志采集服务已停止");
        super.onDestroy();
    }

    @Override
    public void onTaskRemoved(Intent rootIntent) {
        android.util.Log.d(TAG, "应用任务被移除，但服务继续运行");
        super.onTaskRemoved(rootIntent);
    }

    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    private void startForegroundImmediately() {
        android.util.Log.d(TAG, "开始启动前台服务");

        try {
            Notification notification = createSimpleNotification();
            if (notification != null) {
                startForeground(NOTIFICATION_ID, notification);
                isForeground = true;
                android.util.Log.d(TAG, "前台服务启动成功");
            } else {
                android.util.Log.e(TAG, "创建通知失败，无法启动前台服务");
                startFallbackForeground();
            }
        } catch (SecurityException e) {
            android.util.Log.e(TAG, "启动前台服务权限异常", e);
            startFallbackForeground();
        } catch (Exception e) {
            android.util.Log.e(TAG, "启动前台服务失败", e);
            startFallbackForeground();
        }
    }

    private Notification createSimpleNotification() {
        try {
            NotificationCompat.Builder builder;
            builder = new NotificationCompat.Builder(this, CHANNEL_ID);
            return builder.setContentTitle("日志采集服务")
                    .setContentText("服务启动中...")
                    .setSmallIcon(getNotificationIcon())
                    .setPriority(NotificationCompat.PRIORITY_LOW)
                    .setOngoing(true)
                    .setOnlyAlertOnce(true)
                    .build();
        } catch (Exception e) {
            android.util.Log.e(TAG, "创建简单通知失败", e);
            return null;
        }
    }

    private void startFallbackForeground() {
        try {
            NotificationCompat.Builder builder = new NotificationCompat.Builder(this, CHANNEL_ID);
            Notification notification = builder
                    .setContentTitle("日志服务")
                    .setContentText("运行中")
                    .setSmallIcon(android.R.drawable.ic_dialog_info)
                    .setPriority(NotificationCompat.PRIORITY_LOW)
                    .build();
            startForeground(NOTIFICATION_ID, notification);
            isForeground = true;
            android.util.Log.d(TAG, "备用前台服务启动成功");
        } catch (Exception e) {
            android.util.Log.e(TAG, "备用前台服务也启动失败", e);
        }
    }

    public void createNotificationChannel() {
        try {
            NotificationChannel channel = new NotificationChannel(
                    CHANNEL_ID,
                    CHANNEL_NAME,
                    NotificationManager.IMPORTANCE_LOW
            );
            channel.setDescription("用于采集应用自身运行日志");
            channel.setShowBadge(false);
            channel.setLockscreenVisibility(Notification.VISIBILITY_PRIVATE);

            if (notificationManager != null) {
                notificationManager.createNotificationChannel(channel);
                android.util.Log.d(TAG, "通知渠道创建成功");
            } else {
                android.util.Log.e(TAG, "NotificationManager 为 null");
            }
        } catch (Exception e) {
            android.util.Log.e(TAG, "创建通知渠道失败", e);
        }
    }

    private void updateNotification(String contentText) {
        if (isForeground) {
            mainHandler.post(() -> {
                try {
                    NotificationCompat.Builder builder;
                    builder = new NotificationCompat.Builder(this, CHANNEL_ID);

                    Notification notification = builder
                            .setContentTitle("日志采集服务")
                            .setContentText(contentText)
                            .setSmallIcon(getNotificationIcon())
                            .setPriority(NotificationCompat.PRIORITY_LOW)
                            .setOngoing(true)
                            .setOnlyAlertOnce(true)
                            .build();

                    if (notificationManager != null) {
                        notificationManager.notify(NOTIFICATION_ID, notification);
                    }
                } catch (Exception e) {
                    android.util.Log.e(TAG, "更新通知失败", e);
                }
            });
        }
    }

    private int getNotificationIcon() {
        try {
            int icon = R.mipmap.ic_launcher;
            if (icon == 0) {
                icon = android.R.drawable.ic_dialog_info;
            }
            return icon;
        } catch (Exception e) {
            return android.R.drawable.ic_dialog_info;
        }
    }

    private void startLogCollection() {
        android.util.Log.d(TAG, "准备启动日志收集");

        if (logcatThread != null && logcatThread.isAlive()) {
            android.util.Log.d(TAG, "停止现有日志收集线程");
            logcatThread.interrupt();
            try {
                logcatThread.join(1000);
            } catch (InterruptedException e) {
                android.util.Log.w(TAG, "等待旧线程结束被中断", e);
            }
        }

        updateNotification("正在初始化日志收集...");

        logcatThread = new Thread(new LogCollectorRunnable());
        logcatThread.setName("AppLogCollector-Thread");
        logcatThread.setPriority(Thread.MIN_PRIORITY);
        logcatThread.start();

        android.util.Log.d(TAG, "日志收集线程已启动");
    }

    private void stopLogCollection() {
        android.util.Log.d(TAG, "开始停止日志收集");
        isRunning.set(false);

        updateNotification("正在停止服务...");

        if (logcatThread != null && logcatThread.isAlive()) {
            logcatThread.interrupt();
            try {
                logcatThread.join(2000);
            } catch (InterruptedException e) {
                android.util.Log.w(TAG, "等待日志线程结束被中断", e);
            }
        }

        if (logcatProcess != null) {
            try {
                logcatProcess.destroy();
            } catch (Exception e) {
                android.util.Log.e(TAG, "停止logcat进程失败", e);
            }
            logcatProcess = null;
        }

        closeCurrentWriter();

        if (isForeground) {
            try {
                stopForeground(true);
                isForeground = false;
                android.util.Log.d(TAG, "前台服务已停止");
            } catch (Exception e) {
                android.util.Log.e(TAG, "停止前台服务失败", e);
            }
        }

        android.util.Log.d(TAG, "日志收集已完全停止");
    }

    private List<String> buildLogcatCommand() {
        List<String> command = new ArrayList<>();
        command.add("logcat");
        command.add("-v");
        command.add("time");
        command.add("--pid=" + android.os.Process.myPid());
        command.add("*:V");
        android.util.Log.d(TAG, "Logcat命令: " + command);
        return command;
    }

    private class LogCollectorRunnable implements Runnable {
        @Override
        public void run() {
            android.util.Log.d(TAG, "日志收集线程启动");
            mainHandler.post(() -> updateNotification("正在采集应用日志..."));

            try {
                File logDir = new File(getFilesDir(), LOG_DIR);
                File appLogDir = new File(logDir, APP_LOG_SUBDIR);
                if (!appLogDir.exists() && !appLogDir.mkdirs()) {
                    android.util.Log.e(TAG, "无法创建日志目录: " + appLogDir.getAbsolutePath());
                    mainHandler.post(() -> updateNotification("创建日志目录失败"));
                    return;
                }

                List<String> command = buildLogcatCommand();
                android.util.Log.d(TAG, "执行logcat命令: " + command);

                mainHandler.post(() -> updateNotification("正在启动logcat进程..."));

                ProcessBuilder processBuilder = new ProcessBuilder(command);
                processBuilder.redirectErrorStream(true);
                logcatProcess = processBuilder.start();

                BufferedReader reader = new BufferedReader(
                        new InputStreamReader(logcatProcess.getInputStream()));

                currentFile = createNewLogFile(appLogDir);
                currentWriter = new BufferedWriter(new FileWriter(currentFile, true));

                android.util.Log.d(TAG, "开始写入日志文件: " + currentFile.getAbsolutePath());
                mainHandler.post(() -> updateNotification("正在采集日志..."));

                String line;
                int lineCount = 0;
                long lastStatusLogTime = System.currentTimeMillis();
                long lastNotificationUpdate = System.currentTimeMillis();
                long lastFileCheckTime = System.currentTimeMillis();
                final long FILE_CHECK_INTERVAL = 5000;

                while (isRunning.get() && !Thread.currentThread().isInterrupted()) {
                    try {
                        long currentTime = System.currentTimeMillis();
                        if (currentTime - lastFileCheckTime > FILE_CHECK_INTERVAL) {
                            if (currentFile != null && !currentFile.exists()) {
                                android.util.Log.w(TAG, "当前日志文件已被删除，重新创建新文件");
                                mainHandler.post(() -> updateNotification("检测到文件被删除，重新创建..."));

                                closeCurrentWriter();
                                currentFile = createNewLogFile(appLogDir);
                                currentWriter = new BufferedWriter(new FileWriter(currentFile, true));

                                android.util.Log.d(TAG, "已创建新日志文件: " + currentFile.getAbsolutePath());
                                mainHandler.post(() -> updateNotification("已重新创建日志文件"));
                            }
                            lastFileCheckTime = currentTime;
                        }

                        line = reader.readLine();
                        if (line != null) {
                            String enhancedLine = enhanceLogLine(line);

                            if (currentWriter == null || (currentFile != null && !currentFile.exists())) {
                                android.util.Log.w(TAG, "日志文件状态异常，重新初始化");
                                closeCurrentWriter();
                                currentFile = createNewLogFile(appLogDir);
                                currentWriter = new BufferedWriter(new FileWriter(currentFile, true));
                            }

                            try {
                                currentWriter.write(enhancedLine);
                                currentWriter.newLine();
                                currentWriter.flush();
                            } catch (IOException e) {
                                if (e.getMessage() != null &&
                                        (e.getMessage().contains("ENOENT") ||
                                                e.getMessage().contains("No such file") ||
                                                e.getMessage().contains("Stream closed"))) {

                                    android.util.Log.w(TAG, "写入日志失败，文件可能被删除，重新创建: " + e.getMessage());
                                    closeCurrentWriter();
                                    currentFile = createNewLogFile(appLogDir);
                                    currentWriter = new BufferedWriter(new FileWriter(currentFile, true));

                                    currentWriter.write(enhancedLine);
                                    currentWriter.newLine();
                                    currentWriter.flush();
                                } else {
                                    throw e;
                                }
                            }

                            lineCount++;

                            if (lineCount % 100 == 0 || (currentTime - lastStatusLogTime) > 30000) {
                                android.util.Log.d(TAG, "已采集 " + lineCount + " 行日志");
                                lastStatusLogTime = currentTime;
                            }

                            if (currentTime - lastNotificationUpdate > 30000) {
                                int finalLineCount = lineCount;
                                mainHandler.post(() ->
                                        updateNotification("已采集 " + finalLineCount + " 行日志"));
                                lastNotificationUpdate = currentTime;
                            }

                            if (currentFile.length() >= MAX_FILE_SIZE) {
                                android.util.Log.d(TAG, "日志文件达到大小限制，开始轮转");
                                mainHandler.post(() -> updateNotification("正在轮转日志文件..."));
                                rotateLogFile(appLogDir);
                                mainHandler.post(() -> updateNotification("正在采集日志..."));
                            }
                        } else {
                            android.util.Log.d(TAG, "Logcat 流已结束");
                            break;
                        }
                    } catch (IOException e) {
                        if (isRunning.get()) {
                            android.util.Log.e(TAG, "读取日志流失败", e);
                        }
                        break;
                    }
                }

                android.util.Log.d(TAG, "日志采集完成，共采集 " + lineCount + " 行日志");
                mainHandler.post(() -> updateNotification("日志采集已完成"));
            } catch (IOException e) {
                android.util.Log.e(TAG, "启动日志采集失败", e);
                mainHandler.post(() -> updateNotification("日志采集启动失败"));
            } finally {
                android.util.Log.d(TAG, "日志采集线程结束");
                closeCurrentWriter();
                if (logcatProcess != null) {
                    logcatProcess.destroy();
                }
            }
        }

        private String enhanceLogLine(String originalLine) {
            String timestamp = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS", Locale.getDefault())
                    .format(new Date());
            return String.format("[%s] %s", timestamp, originalLine);
        }
    }

    private File createNewLogFile(File logDir) {
        String timestamp = new SimpleDateFormat("yyyyMMdd_HHmmss", Locale.getDefault())
                .format(new Date());
        String fileName = FILE_PREFIX + timestamp + FILE_SUFFIX;
        File newFile = new File(logDir, fileName);
        android.util.Log.d(TAG, "创建新日志文件: " + newFile.getAbsolutePath());
        return newFile;
    }

    private void rotateLogFile(File logDir) {
        closeCurrentWriter();

        currentFile = createNewLogFile(logDir);
        try {
            currentWriter = new BufferedWriter(new FileWriter(currentFile, true));
            android.util.Log.d(TAG, "日志文件轮转完成");
        } catch (IOException e) {
            android.util.Log.e(TAG, "创建新日志文件失败", e);
            return;
        }

        cleanupOldFiles(logDir);
    }

    private void cleanupOldFiles(File logDir) {
        File[] logFiles = logDir.listFiles((dir, name) -> name.startsWith(FILE_PREFIX) && name.endsWith(FILE_SUFFIX));

        if (logFiles != null && logFiles.length > MAX_FILES) {
            Arrays.sort(logFiles, Comparator.comparingLong(File::lastModified));

            int filesToDelete = logFiles.length - MAX_FILES;
            for (int i = 0; i < filesToDelete; i++) {
                if (logFiles[i].delete()) {
                    android.util.Log.d(TAG, "删除旧日志文件: " + logFiles[i].getName());
                } else {
                    android.util.Log.w(TAG, "删除旧日志文件失败: " + logFiles[i].getName());
                }
            }
        }
    }

    private void closeCurrentWriter() {
        if (currentWriter != null) {
            try {
                currentWriter.close();
                android.util.Log.d(TAG, "日志写入器已关闭");
            } catch (IOException e) {
                android.util.Log.e(TAG, "关闭日志写入器失败", e);
            }
            currentWriter = null;
        }
    }
}

File: app/src/main/java/com/qimian233/ztool/service/LogCollectorService.kt
package com.qimian233.ztool.service

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.Service
import android.content.Intent
import android.os.Handler
import android.os.IBinder
import android.os.Looper
import android.util.Log
import androidx.core.app.NotificationCompat
import com.qimian233.ztool.R
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.File
import java.io.FileWriter
import java.io.IOException
import java.io.InputStreamReader
import java.text.SimpleDateFormat
import java.util.Arrays
import java.util.Comparator
import java.util.Date
import java.util.Locale
import java.util.concurrent.atomic.AtomicBoolean

/**
 * App log collection service (no root required; filters logcat by PID)
 */
class LogCollectorService : Service() {
    private var logcatProcess: Process? = null
    private var currentWriter: BufferedWriter? = null
    private var currentFile: File? = null
    private val isRunning = AtomicBoolean(false)
    private var logcatThread: Thread? = null
    private var isRestartMode = false
    private var isForeground = false
    private var mainHandler: Handler? = null
    private var notificationManager: NotificationManager? = null

    override fun onCreate() {
        super.onCreate()
        Log.d(TAG, "service onCreate() start")

        mainHandler = Handler(Looper.getMainLooper())
        notificationManager = getSystemService(NotificationManager::class.java)

        createNotificationChannel()
        startForegroundImmediately()

        Log.d(TAG, "service onCreate() done")
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        Log.d(TAG, "service onStartCommand() start")

        intent?.let {
            isRestartMode = it.getBooleanExtra("is_restart", false)
        }

        if (isRestartMode) {
            Log.d(TAG, "started in service restart mode")
        }

        if (!isForeground) {
            Log.w(TAG, "foreground service not started, starting now")
            startForegroundImmediately()
        }

        if (!isRunning.get()) {
            isRunning.set(true)
            mainHandler?.postDelayed({
                Log.d(TAG, "starting log collection")
                startLogCollection()
            }, 100)
        } else {
            Log.d(TAG, "service already running")
        }

        Log.d(TAG, "service onStartCommand() done")
        return START_STICKY
    }

    override fun onDestroy() {
        Log.d(TAG, "service onDestroy() start")
        stopLogCollection()
        Log.d(TAG, "log collection service stopped")
        super.onDestroy()
    }

    override fun onTaskRemoved(rootIntent: Intent?) {
        Log.d(TAG, "app task removed, service keeps running")
        super.onTaskRemoved(rootIntent)
    }

    override fun onBind(intent: Intent?): IBinder? {
        return null
    }

    private fun startForegroundImmediately() {
        Log.d(TAG, "starting foreground service")

        try {
            val notification = createSimpleNotification()
            if (notification != null) {
                startForeground(NOTIFICATION_ID, notification)
                isForeground = true
                Log.d(TAG, "foreground service started")
            } else {
                Log.e(TAG, "failed to create notification, cannot start foreground service")
                startFallbackForeground()
            }
        } catch (e: SecurityException) {
            Log.e(TAG, "security exception starting foreground service", e)
            startFallbackForeground()
        } catch (e: Exception) {
            Log.e(TAG, "failed to start foreground service", e)
            startFallbackForeground()
        }
    }

    private fun createSimpleNotification(): Notification? {
        return try {
            val builder = NotificationCompat.Builder(this, CHANNEL_ID)
            builder.setContentTitle("日志采集服务")
                .setContentText("日志服务运行中")
                .setSmallIcon(getNotificationIcon())
                .setPriority(NotificationCompat.PRIORITY_LOW)
                .setOngoing(true)
                .setOnlyAlertOnce(true)
                .build()
        } catch (e: Exception) {
            Log.e(TAG, "failed to create simple notification", e)
            null
        }
    }

    private fun startFallbackForeground() {
        try {
            val builder = NotificationCompat.Builder(this, CHANNEL_ID)
            val notification = builder
                .setContentTitle("日志服务")
                .setContentText("运行中")
                .setSmallIcon(android.R.drawable.ic_dialog_info)
                .setPriority(NotificationCompat.PRIORITY_LOW)
                .build()
            startForeground(NOTIFICATION_ID, notification)
            isForeground = true
            Log.d(TAG, "fallback foreground service started")
        } catch (e: Exception) {
            Log.e(TAG, "fallback foreground service also failed to start", e)
        }
    }

    fun createNotificationChannel() {
        try {
            val channel = NotificationChannel(
                CHANNEL_ID,
                CHANNEL_NAME,
                NotificationManager.IMPORTANCE_LOW
            )
            channel.description = "用于采集应用自身运行日志"
            channel.setShowBadge(false)
            channel.lockscreenVisibility = Notification.VISIBILITY_PRIVATE

            val manager = notificationManager
            if (manager != null) {
                manager.createNotificationChannel(channel)
                Log.d(TAG, "notification channel created")
            } else {
                Log.e(TAG, "NotificationManager is null")
            }
        } catch (e: Exception) {
            Log.e(TAG, "failed to create notification channel", e)
        }
    }

    private fun getNotificationIcon(): Int {
        return try {
            val icon = R.mipmap.ic_launcher
            if (icon == 0) {
                android.R.drawable.ic_dialog_info
            } else {
                icon
            }
        } catch (e: Exception) {
            android.R.drawable.ic_dialog_info
        }
    }

    private fun startLogCollection() {
        Log.d(TAG, "preparing to start log collection")

        val existingThread = logcatThread
        if (existingThread != null && existingThread.isAlive) {
            Log.d(TAG, "stopping existing log collection thread")
            existingThread.interrupt()
            try {
                existingThread.join(1000)
            } catch (e: InterruptedException) {
                Log.w(TAG, "interrupted while waiting for old thread to finish", e)
            }
        }

        val collectorThread = Thread(LogCollectorRunnable())
        collectorThread.name = "AppLogCollector-Thread"
        collectorThread.priority = Thread.MIN_PRIORITY
        logcatThread = collectorThread
        collectorThread.start()

        Log.d(TAG, "log collection thread started")
    }

    private fun stopLogCollection() {
        Log.d(TAG, "stopping log collection")
        isRunning.set(false)

        val thread = logcatThread
        if (thread != null && thread.isAlive) {
            thread.interrupt()
            try {
                thread.join(2000)
            } catch (e: InterruptedException) {
                Log.w(TAG, "interrupted while waiting for log thread to finish", e)
            }
        }

        val process = logcatProcess
        if (process != null) {
            try {
                process.destroy()
            } catch (e: Exception) {
                Log.e(TAG, "failed to stop logcat process", e)
            }
            logcatProcess = null
        }

        closeCurrentWriter()

        if (isForeground) {
            try {
                stopForeground(STOP_FOREGROUND_REMOVE)
                isForeground = false
                Log.d(TAG, "foreground service stopped")
            } catch (e: Exception) {
                Log.e(TAG, "failed to stop foreground service", e)
            }
        }

        Log.d(TAG, "log collection fully stopped")
    }

    private fun buildLogcatCommand(): List<String> {
        val command = ArrayList<String>()
        command.add("logcat")
        command.add("-v")
        command.add("time")
        command.add("--pid=" + android.os.Process.myPid())
        command.add("*:V")
        Log.d(TAG, "logcat command: $command")
        return command
    }

    private inner class LogCollectorRunnable : Runnable {
        override fun run() {
            Log.d(TAG, "log collection thread running")

            try {
                val logDir = File(filesDir, LOG_DIR)
                val appLogDir = File(logDir, APP_LOG_SUBDIR)
                if (!appLogDir.exists() && !appLogDir.mkdirs()) {
                    Log.e(TAG, "cannot create log dir: " + appLogDir.absolutePath)
                    return
                }

                val command = buildLogcatCommand()
                Log.d(TAG, "running logcat command: $command")

                val processBuilder = ProcessBuilder(command)
                processBuilder.redirectErrorStream(true)
                val process = processBuilder.start()
                logcatProcess = process

                val reader = BufferedReader(InputStreamReader(process.inputStream))

                val logFile = createNewLogFile(appLogDir)
                currentFile = logFile
                currentWriter = BufferedWriter(FileWriter(logFile, true))

                Log.d(TAG, "writing to log file: " + logFile.absolutePath)

                var line: String?
                var lineCount = 0
                var lastStatusLogTime = System.currentTimeMillis()
                var lastFileCheckTime = System.currentTimeMillis()
                val fileCheckInterval = 5000L

                while (isRunning.get() && !Thread.currentThread().isInterrupted()) {
                    try {
                        val currentTime = System.currentTimeMillis()
                        if (currentTime - lastFileCheckTime > fileCheckInterval) {
                            val checkedFile = currentFile
                            if (checkedFile != null && !checkedFile.exists()) {
                                Log.w(TAG, "current log file was deleted, recreating")

                                closeCurrentWriter()
                                val recreatedFile = createNewLogFile(appLogDir)
                                currentFile = recreatedFile
                                currentWriter = BufferedWriter(FileWriter(recreatedFile, true))

                                Log.d(TAG, "created new log file: " + recreatedFile.absolutePath)
                            }
                            lastFileCheckTime = currentTime
                        }

                        line = reader.readLine()
                        if (line != null) {
                            val enhancedLine = enhanceLogLine(line)

                            val activeFile = currentFile
                            if (currentWriter == null || (activeFile != null && !activeFile.exists())) {
                                Log.w(TAG, "log file state abnormal, reinitializing")
                                closeCurrentWriter()
                                val recreatedFile = createNewLogFile(appLogDir)
                                currentFile = recreatedFile
                                currentWriter = BufferedWriter(FileWriter(recreatedFile, true))
                            }

                            try {
                                currentWriter!!.write(enhancedLine)
                                currentWriter!!.newLine()
                                currentWriter!!.flush()
                            } catch (e: IOException) {
                                if (e.message != null &&
                                    (e.message!!.contains("ENOENT") ||
                                            e.message!!.contains("No such file") ||
                                            e.message!!.contains("Stream closed"))
                                ) {
                                    Log.w(TAG, "log write failed, file may have been deleted, recreating: " + e.message)
                                    closeCurrentWriter()
                                    val recreatedFile = createNewLogFile(appLogDir)
                                    currentFile = recreatedFile
                                    currentWriter = BufferedWriter(FileWriter(recreatedFile, true))

                                    currentWriter!!.write(enhancedLine)
                                    currentWriter!!.newLine()
                                    currentWriter!!.flush()
                                } else {
                                    throw e
                                }
                            }

                            lineCount++

                            if (lineCount % 100 == 0 || (currentTime - lastStatusLogTime) > 30000) {
                                Log.d(TAG, "collected $lineCount lines")
                                lastStatusLogTime = currentTime
                            }

                            val sizedFile = currentFile
                            if (sizedFile != null && sizedFile.length() >= MAX_FILE_SIZE) {
                                Log.d(TAG, "log file reached size limit, rotating")
                                rotateLogFile(appLogDir)
                            }
                        } else {
                            Log.d(TAG, "logcat stream ended")
                            break
                        }
                    } catch (e: IOException) {
                        if (isRunning.get()) {
                            Log.e(TAG, "failed to read log stream", e)
                        }
                        break
                    }
                }

                Log.d(TAG, "collection finished, $lineCount lines collected")
            } catch (e: IOException) {
                Log.e(TAG, "failed to start log collection", e)
            } finally {
                Log.d(TAG, "log collection thread ended")
                closeCurrentWriter()
                logcatProcess?.destroy()
            }
        }

        private fun enhanceLogLine(originalLine: String): String {
            val timestamp = SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS", Locale.getDefault())
                .format(Date())
            return String.format("[%s] %s", timestamp, originalLine)
        }
    }

    private fun createNewLogFile(logDir: File): File {
        val timestamp = SimpleDateFormat("yyyyMMdd_HHmmss", Locale.getDefault())
            .format(Date())
        val fileName = FILE_PREFIX + timestamp + FILE_SUFFIX
        val newFile = File(logDir, fileName)
        Log.d(TAG, "created new log file: " + newFile.absolutePath)
        return newFile
    }

    private fun rotateLogFile(logDir: File) {
        closeCurrentWriter()

        val newFile = createNewLogFile(logDir)
        currentFile = newFile
        try {
            currentWriter = BufferedWriter(FileWriter(newFile, true))
            Log.d(TAG, "log file rotation done")
        } catch (e: IOException) {
            Log.e(TAG, "failed to create new log file", e)
            return
        }

        cleanupOldFiles(logDir)
    }

    private fun cleanupOldFiles(logDir: File) {
        val logFiles = logDir.listFiles { _, name ->
            name.startsWith(FILE_PREFIX) && name.endsWith(FILE_SUFFIX)
        }

        if (logFiles != null && logFiles.size > MAX_FILES) {
            Arrays.sort(logFiles, Comparator.comparingLong { obj: File -> obj.lastModified() })

            val filesToDelete = logFiles.size - MAX_FILES
            for (i in 0 until filesToDelete) {
                if (logFiles[i].delete()) {
                    Log.d(TAG, "deleted old log file: " + logFiles[i].name)
                } else {
                    Log.w(TAG, "failed to delete old log file: " + logFiles[i].name)
                }
            }
        }
    }

    private fun closeCurrentWriter() {
        val writer = currentWriter
        if (writer != null) {
            try {
                writer.close()
                Log.d(TAG, "log writer closed")
            } catch (e: IOException) {
                Log.e(TAG, "failed to close log writer", e)
            }
            currentWriter = null
        }
    }

    companion object {
        private const val TAG = "LogCollectorService"
        private const val NOTIFICATION_ID = 1001
        private const val CHANNEL_ID = "log_collector_channel"
        private const val CHANNEL_NAME = "日志采集服务" // functional: notification channel name (user-facing)

        private const val MAX_FILE_SIZE = 1024L * 1024L // 1MB
        private const val MAX_FILES = 20
        private const val LOG_DIR = "Log"
        private const val APP_LOG_SUBDIR = "app"
        private const val FILE_PREFIX = "app_log_"
        private const val FILE_SUFFIX = ".log"
    }
}
